fix(frontend): fix parsing large workflow graph. Fixes #4179

radcheb opened the pull request:
**Description of your changes:**
Fixes https://github.com/kubeflow/pipelines/issues/4179

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention). 

- [x] Do you want this pull request (PR) cherry-picked into the current release branch?
    
    If yes, use one of the following options:
  
- [x] **(Recommended.)** Ask the PR approver to add the `cherrypick-approved` label to this PR. The release manager adds this PR to the release branch in a batch update.

radcheb commented:
@googlebot I signed it!

Ark-kun commented:
/ok-to-test

Ark-kun commented:
/retest

Bobgy commented on frontend/src/pages/RunDetails.tsx:
```diff
@@ -669,9 +670,18 @@ class RunDetails extends Page<RunDetailsInternalProps, RunDetailsState> {
-      const workflow = JSON.parse(
-        runDetail.pipeline_runtime!.workflow_manifest || '{}',
-      ) as Workflow;
+      const jsonWorkflow = JSON.parse(runDetail.pipeline_runtime!.workflow_manifest || '{}');
```
Thanks! Would it be possible adding a unit test for RunDetails.tsx that goes through this case? So we don't break it accidentally later.

radcheb replied:
Thanks for the review, I will test it.

radcheb replied:
@Bobgy I tested this case in `RunDetails` page and also the `decodeCompressedNodes` in `Utils`. The `kubeflow-pipeline-frontend-test` passed however those tests are now broken: `kubeflow-pipeline-e2e-test`, `kubeflow-pipeline-sample-test` and `kubeflow-pipeline-upgrade-test`.
I am no sure if the reason is related to this PR?

radcheb commented:
/retest

Bobgy commented:
The failed tests had an error building frontend image.

Step #2 - "frontend": Failed to compile.

14886

Step #2 - "frontend": 

14887

Step #2 - "frontend": /src/frontend/src/lib/Utils.tsx

14888

Step #2 - "frontend": TypeScript error in /src/frontend/src/lib/Utils.tsx(379,5):

14889

Step #2 - "frontend": Cannot find name 'reject'. TS2304

14890

Step #2 - "frontend": 

14891

Step #2 - "frontend": 377 | } catch (err) {

14892

Step #2 - "frontend": 378 | logger.error('Error decoding compressedNodes!', err);

14893

Step #2 - "frontend": > 379 | reject(err);

14894

Step #2 - "frontend": | ^

14895

Step #2 - "frontend": 380 | }

14896

Step #2 - "frontend": 381 | }

14897

Step #2 - "frontend": 382 |

radcheb commented:
> The failed tests had an error building frontend image.
> 
> Step #2 - "frontend": Failed to compile.
> 
> 14886
> 
> Step #2 - "frontend":
> 
> 14887
> 
> Step #2 - "frontend": /src/frontend/src/lib/Utils.tsx
> 
> 14888
> 
> Step #2 - "frontend": TypeScript error in /src/frontend/src/lib/Utils.tsx(379,5):
> 
> 14889
> 
> Step #2 - "frontend": Cannot find name 'reject'. TS2304
> 
> 14890
> 
> Step #2 - "frontend":
> 
> 14891
> 
> Step #2 - "frontend": 377 | } catch (err) {
> 
> 14892
> 
> Step #2 - "frontend": 378 | logger.error('Error decoding compressedNodes!', err);
> 
> 14893
> 
> Step #2 - "frontend": > 379 | reject(err);
> 
> 14894
> 
> Step #2 - "frontend": | ^
> 
> 14895
> 
> Step #2 - "frontend": 380 | }
> 
> 14896
> 
> Step #2 - "frontend": 381 | }
> 
> 14897
> 
> Step #2 - "frontend": 382 |

Ok, thanks @Bobgy for your response, I look at it.

radcheb commented:
/retest

radcheb commented:
/test kubeflow-pipeline-frontend-test

Bobgy commented:
@radcheb would you mind reformatting the code?
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubeflow_pipelines/4180/kubeflow-pipeline-frontend-test/1294002360783212544#1:build-log.txt%3A51

radcheb commented:
> @radcheb would you mind reformatting the code?
> https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubeflow_pipelines/4180/kubeflow-pipeline-frontend-test/1294002360783212544#1:build-log.txt%3A51

@Bobgy I rebased on master, reformatted code, and updated snapshot.

Bobgy commented on frontend/src/pages/__snapshots__/RunDetails.test.tsx.snap:
```diff
@@ -1138,6 +1138,12 @@ exports[`RunDetails renders an empty run 1`] = `
+exports[`RunDetails shows a one-node compressed workflow graph 1`] = `
+<div
+  className="page"
+/>
```
sth seems wrong, the snapshot is empty.
If you compare it to snapshot of the one node test, it has some content

Bobgy commented:
@radcheb Thank you for the update, but looks like there is still a problem

Bobgy commented:
/retest

radcheb commented on frontend/src/pages/RunDetails.tsx:
```diff
@@ -677,12 +677,11 @@ class RunDetails extends Page<RunDetailsInternalProps, RunDetailsState> {
-        jsonWorkflow.status.nodes = await decodeCompressedNodes(
-          jsonWorkflow.status.compressedNodes,
-        );
+        jsonWorkflow.status.nodes = decodeCompressedNodes(jsonWorkflow.status.compressedNodes);
```
@Bobgy I am not an expert with React testing but I can't understand why I needed to remove `await` here for tests to pass! If I keep `await` code fails and the page is empty!

Bobgy replied:
Hey, I think there are two problems showing together:
1. the code should have `await` there, but there's sth wrong in the test to make it wait for the gunzip decompress properly
2. when you remove the `await`, the test no longer needs to wait gunzip decompress, so it shows sth in the snapshot, but the snapshot doesn't show any **real** content in there either. That's a problem with our current way for testing this component.

Would you mind waiting a few days to let me refactor this one node graph test case first to solve problem 2? You can mimic my refactored test case to properly test it and figure out what went wrong to problem 1.

Bobgy replied:
Also, these are not specific to react, they are javascript testing problems.

To explain problem 1 a bit more, we use `await flushPromises()` in tests to wait for mocked promises that return immediately, but you didn't mock the zlib library here, it's not clear when it will run the compress callback handler.

For this specific case, I don't see why you cannot use the `gunzipSync` method instead: https://nodejs.org/api/zlib.html#zlib_zlib_gunzipsync_buffer_options
It will avoid your problem.

radcheb replied:
Thanks @Bobgy for your quick response. Actually I can mock also the `decodeCompressedNodes` so we escape the problem with `flushPromises`.

I used async `gunzip` to avoid freezing UI while decoding compressed nodes. Usually, server compress the nodes property when the pipeline is too big, so it may takes some time to decode the base64 and gunzip it.

I don't know if mocking `decodeCompressedNodes` will fix Problem 2, we will see.

radcheb commented on frontend/src/pages/RunDetails.tsx:
```diff
@@ -677,12 +677,11 @@ class RunDetails extends Page<RunDetailsInternalProps, RunDetailsState> {
         delete jsonWorkflow.status.compressedNodes;
       }
       const workflow = jsonWorkflow as Workflow;
+      console.log(JSON.stringify(workflow));
```
TODO: remove it

Bobgy commented:
@radcheb I've never used zlib, just want to double check whether you ran your local instance and verified this PR works with your data in the browser? It seems zlib is a node native module, can it run in the browser?

I'm not familiar with this so I'm asking, but if you tried that there's no problem at all.

radcheb commented:
/retest

radcheb commented:
Hi @Bobgy , `kubeflow-pipeline-sample-test` is broken because of `tensorflow==1.8.0` is no more available for `python 3.7`:
![image](https://user-images.githubusercontent.com/5963615/95100230-317d0180-0731-11eb-8133-1f9c0c007c70.png)

I created this PR https://github.com/kubeflow/pipelines/pull/4582 to fix it. However, `needs-ok-to-test` is needed for tests to run. (I first tried with an empty commit to reproduce the error on another branch but I forget that `needs-ok-to-test` is needed for tests to run).

radcheb commented:
@Bobgy all test are fixed. I ended up mocking `decodeCompressedNodes`.

For testing I added `large-graph-runtime` containing a big workflow to `mock-bakend`. The Run name is `Very large graph`. Without the fix, there is "No Graph to show" in the run page. After the fix the graph shows up successfully:
![image](https://user-images.githubusercontent.com/5963615/95578602-70ae9980-0a34-11eb-9b88-73af60936932.png)

I did the test on chrome on Mac OS X and Linux Ubuntu.

radcheb commented:
@Bobgy regarding your question about **zlib**, I confirm it works in the browser since:
* The local test worked in my mac and linux Chrome
* [Axios](https://github.com/axios/axios), a widely used HTTP client library in js, [is already using zlib](https://github.com/axios/axios/search?q=zlib):
![image](https://user-images.githubusercontent.com/5963615/95578890-fa5e6700-0a34-11eb-83c8-9234d9daa5e1.png)

And they claims [it's supported on almost all browsers](https://github.com/axios/axios#browser-support):
![image](https://user-images.githubusercontent.com/5963615/95578969-106c2780-0a35-11eb-9422-b83147cc5319.png)

Bobgy commented:
Thank you @radcheb, great progress!
I'm just back from vacation, taking a look now.

Bobgy commented on frontend/src/lib/Utils.tsx:
```diff
@@ -358,3 +359,24 @@ export function buildQuery(queriesMap: { [key: string]: string | number | undefi
+export async function decodeCompressedNodes(compressedNodes: string): Promise<object> {
+  return new Promise<object>(resolve => {
+    const compressedBuffer = Buffer.from(compressedNodes, 'base64');
+    zlib.gunzip(compressedBuffer, { windowBits: 15 }, (error, result: Buffer) => {
```
nit: what does the `windowBits: 15` mean?

Can we avoid the magic number?

radcheb replied:
Sorry forgot to remove it.

Bobgy commented on frontend/src/lib/Utils.tsx:
```diff
@@ -358,3 +359,24 @@ export function buildQuery(queriesMap: { [key: string]: string | number | undefi
+    zlib.gunzip(compressedBuffer, { windowBits: 15 }, (error, result: Buffer) => {
+      if (error) {
+        logger.error('failed to gunzip data ', error);
+        resolve(undefined);
```
nit: a more common practice is to reject when there's an error, but it's up to you whether you want to change

radcheb replied:
You're right, I will reject here and catch error in RunDetails.

Bobgy commented on frontend/src/pages/RunDetails.test.tsx:
```diff
@@ -167,6 +168,9 @@ describe('RunDetails', () => {
     warnSpy = jest.spyOn(Utils.logger, 'warn').mockImplementation();
 
+    // Mock decodeCompressedNodes to avoid dequeuing chained promises
+    decodeCompressedNodesSpy = jest.spyOn(Utils, 'decodeCompressedNodes').mockImplementation();
```
nit: `decodeCompressedNodes` itself is very important to test in this test case, because otherwise we are not verifying this feature end to end.

I understand you may be blocked by sth. You can put a TODO comment here too if you cannot figure out by yourself.

If you'd want to try testing it. Here are my ideas:
I'm not sure how `zlib` will trigger the callback, I guess if simply await is not enough, we can use `setTimeout` to explicitly wait some long time period so `zlib` can decompress.

radcheb replied:
Well, I already unit tested `decodeCompressedNodes` in  [Utils.test.tsx](https://github.com/kubeflow/pipelines/pull/4180/files#diff-3c4323c8f2a90f3a3b926a319d85baf1R267). 
The `setTimeout` trick will work for sure but not so elegant.

radcheb replied:
I test the wait and it works, thanks for the hint: 8491c5c6104fcaee170d1b0b6246d654dd7a25b8

Bobgy commented on frontend/src/pages/RunDetails.test.tsx:
```diff
@@ -654,6 +658,35 @@ describe('RunDetails', () => {
+
+    await TestUtils.flushPromises();
+
+    expect(getByTestId('graph')).toMatchInlineSnapshot(`
+      <pre
+        data-testid="graph"
+      >
+        Node node1
+        Node node1-running-placeholder
+        Edge node1 to node1-running-placeholder
+      </pre>
+    `);
```
nice! now it's showing readable content

Bobgy reviewed (commented):
Thank you! all looking good at high level
/lgtm

Read through code thoroughly and only left some nit picking

axelborja commented:
Thanks @radcheb for this fix !! 👏

radcheb commented:
@Bobgy could you review and merge if it's ok ?

Bobgy commented:
> @Bobgy could you review and merge if it's ok ?

@radcheb Sure, I'll review tomorrow.

Bobgy commented on frontend/src/pages/RunDetails.test.tsx:
```diff
@@ -654,6 +653,50 @@ describe('RunDetails', () => {
+    await new Promise(resolve => setTimeout(resolve, 500));
+    jest.useFakeTimers();
+
+    expect(tree).toMatchSnapshot('');
```
this should be
```
expect(getByTestId('graph')).toMatchInlineSnapshot()
```

Bobgy replied:
actually, I'm not sure whether `graph` will actually render, if not you can also just assert something like
```
expect(getAllByTestId('graph')).toEqual([])
```

radcheb replied:
Yes in this case the graph should be empty, I will use your second suggestion.

radcheb replied:
Well `getAllByTestId` didn't work since it requires at least one element in the list, it was rather `queryAllByTestId` who did the job.

Bobgy reviewed (changes requested):
Thank you @radcheb for your perseverance!
This looks awesome!

There's one last minor mistake to fix, otherwise all good!
/lgtm

Bobgy commented:
/lgtm
/approve

Really valuable fix!
@racheb Thank you for your contribution!

radcheb commented:
Thanks @Bobgy for accepting and merging. Will this be included in a future release ?

Bobgy commented:
It's not clear whether we will release 1.0.4 or not. If not, it will come in 1.1.0

Bobgy commented:
Hi @radcheb, we just released 1.0.4 with your fix!

axelborja commented:
Thanks 🙏

radcheb commented:
> Hi @radcheb, we just released 1.0.4 with your fix!

Thanks @Bobgy for your assistance and thus this great news.
